### ISSUE TITLE
Pyright hangs with version 3.11 code base

### ISSUE THREAD

**Issue by flying-sheep** [microsoft/pyright#4353]
**Describe the bug**
Running pyright at my code base hangs.

**To Reproduce**
```bash
$ git clone https://github.com/flying-sheep/fine-coverage
$ cd fine-coverage
$ git checkout 16c22fb5549e058ce55a2c20fa06a3c2362f4493
$ npx --silent --yes pyright@1.1.284 --pythonversion=3.11
No configuration file found.
pyproject.toml file found at /home/phil/Dev/Python/Packaging, Docs/fine-coverage.
Loading pyproject.toml file at /home/phil/Dev/Python/Packaging, Docs/fine-coverage/pyproject.toml
Assuming Python platform Linux
Auto-excluding **/node_modules
Auto-excluding **/__pycache__
Auto-excluding **/.*
stubPath /home/phil/Dev/Python/Packaging, Docs/fine-coverage/typings is not a valid directory.
Searching for source files
Found 7 source files
```

**Expected behavior**
It runs just as it does when not specifying `--pythonversion=3.11` (which causes it to use 3.10 on my system)

**Comment by ifuatgucluer** [comment]
thanks


Philipp A. ***@***.***>, 18 Ara 2022 Paz, 17:19 tarihinde
şunu yazdı:

**Comment by erictraut** [comment]
Thanks for the bug report. The issue was related to the fact that you're using a recursive protocol (`TraceFunction`, which includes a method that returns a `Self | None`) and passing it to `sys.settrace`, which was recently changed to use a recursive type alias called `TraceFunction`. When determining type compatibility between the recursive protocol and the recursive type alias, pyright was recursing deeply down both types, and this resulted in an exponential increase in evaluation time. It would eventually complete, but it would take a long time. 

I've added better recursion detection to prevent this from happening. This will be addressed in the next release.

**Comment by flying-sheep** [comment]
That’s great! Thanks for figuring it out!

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.285, which I just published. It will also be included in a future release of pylance.
